File: python_backend/game_logic/pathsInfo/top/get_pathdata_dict.py
from game_logic.pathsInfo.get_path_dir import _get_path_dir
from game_logic.pathsInfo.PathData import PathData
from game_logic.pathsInfo.get_path_data import _get_path_data
from game_logic.step_funcs.step_funcs import step_func_list
from game_logic.printers.print_board import print_board
from game_logic.test_objects.get_standard_piece_defs import get_standard_piece_defs
from game_logic.test_objects.get_standard_id_dict import get_standard_id_dict
from game_logic.test_objects.sample_board_dicts import sample_board_dicts


def get_pathdata_dict(board, sqr, color, piece_defs, id_dict):
    """get data about every path stemming from sqr, and record in dict"""

    pd_dict = {}
    for step_func in step_func_list:
        # get direction of path
        path_dir = _get_path_dir(step_func)

        # initialize class for given direction
        pd_dict[path_dir] = PathData(path_dir)

        # get path data for given direction
        path_data = _get_path_data(board, sqr, color, piece_defs, id_dict, step_func)

        # initialize class attributes to path data results
        pd_dict[path_dir].coord_path = path_data[0]
        pd_dict[path_dir].path = path_data[1]
        pd_dict[path_dir].pieces = path_data[2]
        pd_dict[path_dir].statuses = path_data[3]
        pd_dict[path_dir].piece_matches_path = path_data[4]

    return pd_dict


if __name__ == "__main__":

    board = sample_board_dicts['check_example3']

    piece_defs = get_standard_piece_defs()
    id_dict = get_standard_id_dict()

    # test 1:
    pd_dict = get_pathdata_dict(board, (4, 4), 'W', piece_defs, id_dict)
    print('test 1:')
    for path in pd_dict.values():
        print_board(board, heading=str(path), highlights=path.coord_path, highlights3=[(4, 4)])
        print("")

File: node.js/pathsInfo/top/getpathDataDict.js
var  getPathDir  = require("../getPathDir");
var  PathData  = require("../PathData");
var  getPathData  = require("../getPathData");
var  stepFuncList  = require("../../helpers/stepFuncs").stepFuncList;
var  getStandardPieceDefs  = require("../../testObjects/getStandardPieceDefs");
var  getStandardIdDict  = require("../../testObjects/getStandardIdDict");
var  sampleBoardDicts  = require("../../testObjects/sampleBoardDicts");
const getPiecesOnPath = require("../getPiecesOnPath");
const mapListXyToRf = require("../../coordType/mapListXyToRf");

/** */
function getPathdataDict(board, sqr, color, pieceDefs, idDict) {
    /*get data about every path stemming from sqr, &&  dict.includes(record)**/
    
    let pdDict = {}
    for (var stepFunc of stepFuncList) {
        
        // get direction of path
        let pathDir = getPathDir(stepFunc)
        
        // initialize class for given direction
        pdDict[pathDir] = new PathData(pathDir)
        
        // get path data for given direction
        let pathData = getPathData(board, sqr, color, pieceDefs, idDict, stepFunc)
        
        //initialize class attributes to path data results
        pdDict[pathDir].coordPath = pathData[0]
        pdDict[pathDir].path = pathData[1]
        pdDict[pathDir].pieces = pathData[2]
        pdDict[pathDir].statuses = pathData[3]
        pdDict[pathDir].pieceMatchesPath = pathData[4]
    }

    return pdDict
}

module.exports = getPathdataDict;

if (require.main === module) {

    const getPieceLocs = require("../../helpers/getPieceLocsOnPath");
    const getMatchesPathLists = require("../../helpers/getMatchesPathLists")
    const getFandR = require("../../helpers/getFandR");
    const printBoard  = require("../../printers/printBoard");
    const getEmptySqrsOnPath = require("../../helpers/getEmptySqrsOnPath")

    let board = sampleBoardDicts['check_example3']
    let pieceDefs = getStandardPieceDefs()
    let idDict = getStandardIdDict()
    var coordPath;

    let pdDict = getPathdataDict(board, [4, 4], 'W', pieceDefs, idDict)
    for (var angle of Object.keys(pdDict)) {
        console.log(`---------------- ${angle} ----------------`.blue)
        coordPath = mapListXyToRf(pdDict[angle].coordPath)
        let pieceLocs = getPieceLocs(board, coordPath)
        var [matchesPath, doesNotMatchPath] = getMatchesPathLists(pieceLocs, pdDict[angle].pieceMatchesPath)
        var [friends, enemies] = getFandR(pieceLocs, pdDict[angle].statuses)
        var emptySqrs = getEmptySqrsOnPath(coordPath, pieceLocs)
        printBoard(board, "friends: green, enemies: red, start square: blue", enemies, friends, ["d4"], emptySqrs)
        printBoard(board, "piece ability matches path: green, does not match path: red, start square: blue", doesNotMatchPath, matchesPath, ["d4"], emptySqrs)    
    }
}